### PR TITLE
Add Ed25519 support to JWT

### PR DESCRIPTION
**Pull request by pwolanin**
Comparing to other PHP JWT libraries this is an important missing feature, and fairly easy to implement

e.g. see: https://github.com/lcobucci/jwt/blob/4.2.x/src/Signer/Eddsa.php

This can be suggested for people on old versions of php or without libsodium (default in php 7.2+)
https://github.com/paragonie/sodium_compat

### CODE REVIEW

**Comment by pwolanin** [comment]
@googlebot I signed it

**Comment by bshaffer** [comment]
@pwolanin this is great, thanks for adding this functionality!

Are you willing to write some tests? If not I'll try to find time to add some, but we definitely need tests before merging this PR.

**Comment by pwolanin** [comment]
@bshaffer can you approve the workflows so the tests run here?

Yes, I can add some tests, but I'd like to be sure they will pass here

**Comment by bshaffer** [comment]
@pwolanin I can approve to run the workflow once you've added tests. You haven't added any, so there's no reason to run them at this time...

**Comment by pwolanin** [comment]
Pushed a first test and start on doc.  I expect when this runs it will fail on older PHP versions that will need the shim installed

**Review by bshaffer — APPROVED** [review]

**Comment by bshaffer** [comment]
Fixed the test error in https://github.com/firebase/php-jwt/pull/344

File: src/JWT.php [hunk 1]
```diff
@@ -216,15 +216,14 @@ public static function sign($msg, $key, $alg = 'HS256')
...
-                    }
-                    if ($alg === 'ES384') {
-                        $signature = self::signatureFromDER($signature, 384);
-                    }
-                    return $signature;
                 }
+                if ($alg === 'ES256') {
+                    $signature = self::signatureFromDER($signature, 256);
+                }
+                if ($alg === 'ES384') {
```
Comment (bshaffer) [inline src/JWT.php:223]: Why not `elseif` here?

File: src/JWT.php [hunk 2]
```diff
@@ -214,14 +216,25 @@ public static function sign($msg, $key, $alg = 'HS256')
...
-                    }
-                    if ($alg === 'ES384') {
-                        $signature = self::signatureFromDER($signature, 384);
-                    }
-                    return $signature;
+                }
+                if ($alg === 'ES256') {
+                    $signature = self::signatureFromDER($signature, 256);
+                }
+                if ($alg === 'ES384') {
```
Comment (bshaffer) [inline src/JWT.php:223]:
```suggestion
                } elseif ($alg === 'ES384') {
```

Comment (pwolanin) [reply src/JWT.php:223]: I just moved the existing code out of the unneeded else {} , but sure.

File: README.md [hunk 3]
```diff
@@ -16,6 +16,13 @@ Use composer to manage your dependencies and download PHP-JWT:
 composer require firebase/php-jwt
 ```
 
+Optionally, install the paragonie/sodium_compat package from composer if your
+php is < 7.2 or does not have libsodium installed:
```
Comment (bshaffer) [inline README.md:20]: In our tests, the `paragonie/sodium_compat` package was only required for 5.6. Is it possible this should actually be "php is < 7.0" ?

File: README.md [hunk 4]
```diff
@@ -16,6 +16,13 @@ Use composer to manage your dependencies and download PHP-JWT:
 composer require firebase/php-jwt
 ```
 
+Optionally, install the paragonie/sodium_compat package from composer if your
```
Comment (bshaffer) [inline README.md:19]:
```suggestion
Optionally, install the `paragonie/sodium_compat` package from composer if your
```

File: README.md [hunk 5]
```diff
@@ -144,6 +151,36 @@ $decoded = JWT::decode($jwt, $publicKey, array('RS256'));
 echo "Decode:\n" . print_r((array) $decoded, true) . "\n";
 ```
 
+Example with EdDSA (libsodium and Ed25519 signature)
+----------------------------
+```php
+use Firebase\JWT\JWT;
+
+// Public and private keys are expected to be Base 64 encoded. The last
```
Comment (bshaffer) [inline README.md:159]:
```suggestion
// Public and private keys are expected to be Base64 encoded. The last
```

File: src/JWT.php [hunk 6]
```diff
@@ -214,14 +216,24 @@ public static function sign($msg, $key, $alg = 'HS256')
...
+                if (!function_exists('sodium_crypto_sign_detached')) {
+                    throw new DomainException('libsodium is not available');
+                }
+                try {
+                    // The last non-empty line is used as the key.
+                    $lines = array_filter(explode("\n", $key));
+                    $key = base64_decode(end($lines));
+                    return sodium_crypto_sign_detached($msg, $key);
+                } catch (Exception $e) {
+                    throw new DomainException($e->getMessage(), 0, $e);
```
Comment (bshaffer) [inline src/JWT.php:236]: Is there a reason (other than staying consistent with the other exceptions thrown from this method) that we are wrapping the libsodium exceptions in `DomainException`?

Comment (pwolanin) [reply src/JWT.php:236]: Just for consistency and so we are not adding new exception types for BC. Also the other exception types may not be defined in  PHP without the library or shim present, so anyone trying to write generic try/catch code will have a harder time.

File: README.md (hunk 3 above)
Comment (pwolanin) [reply README.md:20]: it was only added by default in 7.2:
https://www.php.net/manual/en/sodium.installation.php

The php packages for 7.0 and 7.1 you are using may have had it also added

**Pull request merged**